### ISSUE TITLE
Look into using pytest-xdist to push code to external servers to speed up local tests

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#6605]
I just ran `tox` tests locally for the first time in a while (as it's the first time I've written a significant PR in a while) and they were frustratingly slow. `pytest-xdist` which we already use for running our tests in parallel has the ability to push the code to a remote server and run tests there which could significantly speed up how long it takes to run tests locally. This functionality is described at https://github.com/pytest-dev/pytest-xdist#sending-tests-to-remote-ssh-accounts.

I haven't deeply looked into this and there may be reasons it won't work or is a bad idea, but I think it's probably worth playing with.

EFF has some servers we could probably use for this. Unfortunately, we wouldn't be able to give external developers access to these servers and any options specifically related to them probably shouldn't be merged into this repo, however, there may be some generic config/options/scripts that external developers could easily use to set this up themselves on their own servers.

**Comment by ohemorange** [comment]
Yeah they take a while, not sure they're so much faster on the remote server tbh? I usually only run one or two locally (py27, lint) anyway.

**Comment by ohemorange** [comment]
Personally, both 6 minutes and 10 minutes are over the threshold for "find something else to do while waiting" so I end up switching tasks anyway. If we could get it down to, like, 2 minutes maybe it'd be worth it for me.

**Comment by adferrand** [comment]
I did some benchmark on Ubuntu. For the serie of tox env you mentionned @bmw. After warm up and, servers are closed to me (Germany), 

My machine, a 3 years old gaming laptop: 5 min
Standard Digital Ocean droplet (40$/month): 8 min
CPU optimized Digital Ocean droplet (80$/month): 6 min
High-end CPU optimized Digital Ocean droplet (160$/month): 5 min

**Comment by bmw** [comment]
> If we could get it down to, like, 2 minutes maybe it'd be worth it for me.

Not quite. Running `time tox -e py27,lint` on the server was 2m33.492s.

Another thing we could do that might speed things up both in this case and for all the environments I linked earlier is to run tests and lint at the same time like we do in josepy. We might have to switch from pylint to flake8 first, but I've personally been wanting to do that anyway.

This would result in fewer environments needing to be set up, but that single environment would run more tests. I think this is probably worth doing though, especially if there's an easy way for people to specify to `tox` that they just want to run unit tests or the linter without any of the other tests.

**Comment by bmw** [comment]
There wasn't as much interest in this as I expected either because the time savings wasn't enough or the external machine you'd need to make tests run faster than they currently do isn't feasible for external contributors.

Because of this, I'm closing the PR so we can skip adding this complexity to our test setup.
